src/KNXClient.ts

```diff
@@ -0,0 +1,1561 @@
+      // TCP
+      const timeoutError = new Error(
+        `Connection timeout to ${this._peerHost}:${this._peerPort}`
+      );
+      this._clientSocket.connect(this._peerPort, this._peerHost, () => {
+        // this._timer = setTimeout(() => {
+        //     this._timer = null;
+        //     this.emit(KNXClientEvents.error, timeoutError);
+        // }, 1000 * KNX_CONSTANTS.CONNECT_REQUEST_TIMEOUT);
+        this._awaitingResponseType = KNX_CONSTANTS.CONNECT_RESPONSE;
+        this._clientTunnelSeqNumber = 0;
+        if (this._options.isSecureKNXEnabled)
+          this._sendSecureSessionRequestMessage(
+            new TunnelCRI.TunnelCRI(knxLayer)
+          );
+      });
+    } else {
+      // Multicast
+      this._connectionState = STATE.CONNECTED;
+
+      // 16/03/2022 These two are referring to tunneling connection, but i set it here as well. Non si sa mai.
+      this._numFailedTelegramACK = 0; // 25/12/2021 Reset the failed ACK counter
+      this._clearToSend = true; // 26/12/2021 allow to send
+
+      this._clientTunnelSeqNumber = -1;
+      try {
+        this.emit(KNXClientEvents.connected, this._options);
+      } catch (error) {}
+    }
+  }
+
+  getConnectionStatus() {
+    if (this._clientSocket == null) {
+      throw new Error("No client socket defined");
+    }
+    const timeoutError = new Error(
+      `HeartBeat failure with ${this._peerHost}:${this._peerPort}`
+    );
+    const deadError = new Error(
+      `Connection dead with ${this._peerHost}:${this._peerPort}`
+    );
+    this._heartbeatTimer = setTimeout(() => {
+      this._heartbeatTimer = null;
+      try {
+        if (this.sysLogger !== undefined && this.sysLogger !== null)
+          this.sysLogger.error(
+            "KNXClient: getConnectionStatus Timeout " +
+              this._heartbeatFailures +
+              " out of " +
+              this.max_HeartbeatFailures
+          );
+        // this.emit(KNXClientEvents.error, timeoutError);
+      } catch (error) {}
+
+      this._heartbeatFailures++;
+      if (this._heartbeatFailures >= this.max_HeartbeatFailures) {
+        this._heartbeatFailures = 0;
+        try {
+          this.emit(KNXClientEvents.error, deadError);
+        } catch (error) {}
+        this._setDisconnected(deadError.message);
+      }
+    }, 1000 * KNX_CONSTANTS.CONNECTIONSTATE_REQUEST_TIMEOUT);
+    this._awaitingResponseType = KNX_CONSTANTS.CONNECTIONSTATE_RESPONSE;
+    try {
+      this._sendConnectionStateRequestMessage(this._channelID);
+    } catch (error) {}
+  }
+
+  private async closeSocket() {
+    return new Promise<void>((resolve, reject) => {
+      const cb = (error?) => {
+        if (error) {
+          reject(error);
+        } else {
+          resolve();
+        }
+      };
+
+      if (this._clientSocket instanceof UFPSocket) {
+        this._clientSocket.close(cb);
+      } else {
+        this._clientSocket.end(cb);
+      }
+    });
+  }
```

Comment:
Ho creato questo metodo per gestire la chiusura della socket in base al tipo di socket, è asincrono mentre Disconnect attualmente è sincrono, secondo me bisognerebbe far diventare Disconnect asincrono ma vorrei che mi dicessi anche tu cosa ne pensi.

Reply:
Inoltre ci aggiungerei un `this._clientSocket = null` e farei una verifica se la socket esiste prima di chiuderla per evitare errori di doppia chiusura della socket come nel caso menzionato di seguito

Reply:
La disconnessione è in realtà un processo sincrono e ti spiego perchè:

1. Decido di disconnettermi
2. Invio una DISCONNECT_REQUEST dal nodo al gateway KNX/IP
3. Aspetto (mi sembra di ricordare circa 10 secondi) che il gateway KNX/IP mi mandi l'ACKNOWLEDGE alla disconnessione.
5. Se arriva, bene, chiudo il socked, altrimenti chiudo il socket ugualmente e amen.
Questo se la connessione è in TunnelUDP. Se, invece è Multicast, chiudo il socket subito.

Devo quindi aspettare tutte queste cose prima di poter dare l'ok al client che è avvenuta la disconnessione.
E' per questo che l'ho fatta sincrona, anche perchè non ho impostato nessun Callback.

Reply:
Questo è chiaro ma a livello codice tu se non attendi che close/end ritornino non puoi essere sicuro che la socket sia chiusa